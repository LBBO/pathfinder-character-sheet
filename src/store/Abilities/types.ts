export type abilityName = 'strength' | 'dexterity' | 'constitution' | 'intelligence' | 'wisdom' | 'charisma'

export type AbilityAttributes = {
  score: number
  modifier: number
  temporaryAdjustment?: number
  temporaryModifier?: number
}

export type AbilityState = {
  [k in abilityName]: AbilityAttributes
}

export const SET_ABILITY_SCORE = 'SET_ABILITY_SCORE'
export const SET_ABILITY_TEMP_ADJUSTMENT = 'SET_ABILITY_TEMP_ADJUSTMENT'

export type SetAbilityScoreAction = {
  type: typeof SET_ABILITY_SCORE
  ability: keyof AbilityState
  score: number
}

export type SetAbilityTempAdjustmentAction = {
  type: typeof SET_ABILITY_TEMP_ADJUSTMENT
  ability: keyof AbilityState
  adjustment: number
}

export type AbilitiesActionTypes = SetAbilityTempAdjustmentAction | SetAbilityScoreAction
