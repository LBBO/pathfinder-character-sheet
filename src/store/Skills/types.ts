export const quickSkillDefinitions = {
  acrobatics: {
    baseAbility: "dexterity",
    isTrainedOnly: false
  },
  appraise: {
    baseAbility: "intelligence",
    isTrainedOnly: false
  },
  bluff: {
    baseAbility: "charisma",
    isTrainedOnly: false
  },
  climb: {
    baseAbility: "strength",
    isTrainedOnly: false
  },
  craft: {
    baseAbility: "intelligence",
    isTrainedOnly: false
  },
  diplomacy: {
    baseAbility: "charisma",
    isTrainedOnly: false
  },
  disableDevice: {
    baseAbility: "dexterity",
    isTrainedOnly: true
  },
  disguise: {
    baseAbility: "charisma",
    isTrainedOnly: false
  },
  escapeArtist: {
    baseAbility: "dexterity",
    isTrainedOnly: false
  },
  fly: {
    baseAbility: "dexterity",
    isTrainedOnly: false
  },
  handleAnimal: {
    baseAbility: "charisma",
    isTrainedOnly: true
  },
  heal: {
    baseAbility: "wisdom",
    isTrainedOnly: false
  },
  intimidate: {
    baseAbility: "charisma",
    isTrainedOnly: false
  },
  knowledgeArcana: {
    baseAbility: "intelligence",
    isTrainedOnly: true
  },
  knowledgeDungeoneering: {
    baseAbility: "intelligence",
    isTrainedOnly: true
  },
  knowledgeEngineering: {
    baseAbility: "intelligence",
    isTrainedOnly: true
  },
  knowledgeGeography: {
    baseAbility: "intelligence",
    isTrainedOnly: true
  },
  knowledgeHistory: {
    baseAbility: "intelligence",
    isTrainedOnly: true
  },
  knowledgeLocal: {
    baseAbility: "intelligence",
    isTrainedOnly: true
  },
  knowledgeNature: {
    baseAbility: "intelligence",
    isTrainedOnly: true
  },
  knowledgeNobility: {
    baseAbility: "intelligence",
    isTrainedOnly: true
  },
  knowledgePlanes: {
    baseAbility: "intelligence",
    isTrainedOnly: true
  },
  knowledgeReligion: {
    baseAbility: "intelligence",
    isTrainedOnly: true
  },
  linguistics: {
    baseAbility: "intelligence",
    isTrainedOnly: true
  },
  perception: {
    baseAbility: "wisdom",
    isTrainedOnly: false
  },
  perform: {
    baseAbility: "charisma",
    isTrainedOnly: false
  },
  profession: {
    baseAbility: "wisdom",
    isTrainedOnly: true
  },
  ride: {
    baseAbility: "dexterity",
    isTrainedOnly: false
  },
  senseMotive: {
    baseAbility: "wisdom",
    isTrainedOnly: false
  },
  sleightOfHand: {
    baseAbility: "dexterity",
    isTrainedOnly: true
  },
  spellcraft: {
    baseAbility: "intelligence",
    isTrainedOnly: true
  },
  stealth: {
    baseAbility: "dexterity",
    isTrainedOnly: false
  },
  survival: {
    baseAbility: "wisdom",
    isTrainedOnly: false
  },
  swim: {
    baseAbility: "strength",
    isTrainedOnly: false
  },
  useMagicDevice: {
    baseAbility: "charisma",
    isTrainedOnly: true
  }
}

export type Skill = {
  name: string,
  abilityModifier: number,
  ranks: number,
  miscModifier: number,
  totalBonus: number,
  isClassSkill: boolean
}

export type SkillState = {
  [k in keyof typeof quickSkillDefinitions]: Skill
}

export const UPDATE_SKILL_RANKS = 'UPDATE_SKILL_RANKS'
export const UPDATE_SKILL_MISC_MODIFIER = 'UPDATE_SKILL_MISC_MODIFIER'

export type UpdateSkillRankAction = {
  type: typeof UPDATE_SKILL_RANKS,
  payload: {
    skillName: keyof typeof quickSkillDefinitions,
    ranks: number,
  }
}

export type UpdateSkillMiscModifierAction = {
  type: typeof UPDATE_SKILL_MISC_MODIFIER,
  payload: {
    skillName: keyof typeof quickSkillDefinitions,
    miscModifier: number,
  }
}

export type UpdateSkillActionTypes = UpdateSkillRankAction | UpdateSkillMiscModifierAction
