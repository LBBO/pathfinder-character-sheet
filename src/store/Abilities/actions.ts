import {
  abilityName,
  SET_ABILITY_SCORE,
  SET_ABILITY_TEMP_ADJUSTMENT,
  SetAbilityScoreAction,
  SetAbilityTempAdjustmentAction,
} from './types'

export const setAbilityScore = (ability: abilityName, newScore: number): SetAbilityScoreAction => {
  if (newScore >= 0 && newScore <= 20) {
    return {
      type: SET_ABILITY_SCORE,
      ability: ability,
      score: newScore,
    }
  } else {
    throw new Error('Cant set ability value higher than 20 or lower than 0')
  }
}

export const setAbilityTempAdjustment = (ability: abilityName, newAdjustment: number): SetAbilityTempAdjustmentAction => {
  if (newAdjustment >= 0 && newAdjustment <= 20) {
    return {
      type: SET_ABILITY_TEMP_ADJUSTMENT,
      ability: ability,
      adjustment: newAdjustment,
    }
  } else {
    throw new Error('Cant set ability value higher than 20 or lower than 0')
  }
}
