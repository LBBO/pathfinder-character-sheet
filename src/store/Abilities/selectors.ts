import { AbilityAttributes, abilityName, AbilityState } from './types'

export type AbilityModifiers = {
  [k in abilityName]: number
}

export const getModifierFromScore = (score: number) => Math.floor(score / 2) - 5

export const getCurrentModifiersFromAbilities = (abilitiesState: AbilityState) => (Object.entries(abilitiesState) as Array<[abilityName, AbilityAttributes]>)
  .reduce((modifiers: AbilityModifiers, [abilityName, values]) => {
    modifiers[abilityName] = getModifierFromScore(values.score + values.temporaryAdjustment)

    return modifiers
  }, {} as AbilityModifiers)
