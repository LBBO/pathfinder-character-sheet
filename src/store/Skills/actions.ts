import {
  SkillName,
  UPDATE_SKILL_IS_CLASS_SKILL,
  UPDATE_SKILL_MISC_MODIFIER,
  UPDATE_SKILL_RANKS,
  UpdateSkillActionTypes,
} from './types'

export const setSkillRanks = (skillName: SkillName, ranks: number): UpdateSkillActionTypes => {
  return {
    type: UPDATE_SKILL_RANKS,
    payload: {
      skillName,
      ranks,
    },
  }
}

export const setSkillMiscMod = (skillName: SkillName, miscModifier: number): UpdateSkillActionTypes => {
  return {
    type: UPDATE_SKILL_MISC_MODIFIER,
    payload: {
      skillName,
      miscModifier,
    },
  }
}

export const setIsSkillClassSkill = (skillName: SkillName, isClassSkill: boolean): UpdateSkillActionTypes => {
  return {
    type: UPDATE_SKILL_IS_CLASS_SKILL,
    payload: {
      skillName,
      isClassSkill,
    }
  }
}
