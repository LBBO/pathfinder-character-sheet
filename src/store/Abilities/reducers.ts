import { AbilityState } from './types'
import { getModifierFromScore } from './selectors'
import { RootActionTypes } from '../root-reducer'

export const createDefaultAbility = () => ({
  score: 10,
  temporaryAdjustment: 0,
})

const initialState: AbilityState = {
  charisma: createDefaultAbility(),
  constitution: createDefaultAbility(),
  dexterity: createDefaultAbility(),
  intelligence: createDefaultAbility(),
  strength: createDefaultAbility(),
  wisdom: createDefaultAbility(),
}

export const AbilitiesReducer = (state: AbilityState = initialState, action?: RootActionTypes): AbilityState => {
  switch (action?.type) {
    case 'SET_ABILITY_SCORE':
      return {
        ...state,
        [action.ability]: {
          ...state[action.ability],
          // NaN --> 0
          score: action.score || 0,
          modifier: getModifierFromScore(action.score),
        },
      }
    case 'SET_ABILITY_TEMP_ADJUSTMENT':
      return {
        ...state,
        [action.ability]: {
          ...state[action.ability],
          // NaN --> 0
          temporaryAdjustment: action.adjustment || 0,
        },
      }
    default:
      return state
  }
}
