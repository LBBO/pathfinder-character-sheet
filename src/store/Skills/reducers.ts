import {
  quickSkillDefinitions,
  Skill,
  SkillState,
  UPDATE_SKILL_IS_CLASS_SKILL,
  UPDATE_SKILL_MISC_MODIFIER,
  UPDATE_SKILL_RANKS,
  UpdateSkillActionTypes,
} from './types'
import { abilityName } from '../Abilities/types'
import { AbilityModifiers } from '../Abilities/selectors'

export const updateTotalSkillBonus = (skill: Skill) => {
  skill.totalBonus = skill.abilityModifier + skill.ranks + skill.miscModifier + (skill.isClassSkill && skill.ranks >= 1 ? 3 : 0)
}

export const createInitialState = (abilityModifiers: AbilityModifiers) => {
  const result = {} as SkillState

  (Object.keys(quickSkillDefinitions) as Array<keyof SkillState>)
    .forEach((skillName: keyof SkillState) => {
      const baseAbilityName = quickSkillDefinitions[skillName].baseAbility as abilityName
      result[skillName] = {
        name: skillName,
        isClassSkill: false,
        abilityModifier: abilityModifiers[baseAbilityName],
        miscModifier: 0,
        ranks: 0,
        totalBonus: abilityModifiers[baseAbilityName],
      }

      updateTotalSkillBonus(result[skillName])
    })

  return result
}

export const applyNewModifiers = (state: SkillState, abilities: AbilityModifiers): SkillState => {
  const copy = { ...state };

  (Object.entries(state) as Array<[keyof SkillState, Skill]>).forEach(([skillName, skill]) => {
    const baseAbilityName = quickSkillDefinitions[skillName].baseAbility as abilityName

    copy[skillName] = {
      ...skill,
      abilityModifier: abilities[baseAbilityName],
    }

    updateTotalSkillBonus(copy[skillName])
  })

  return copy
}

export const SkillsReducer = (
  state?: SkillState,
  action?: UpdateSkillActionTypes,
  abilities?: AbilityModifiers,
  abilitiesChanged = false
): SkillState => {
  if (!state) {
    return createInitialState(abilities!)
  }

  if (abilitiesChanged) {
    return applyNewModifiers(state, abilities!)
  }

  let updatedSkill: Skill
  switch (action?.type) {
    case UPDATE_SKILL_MISC_MODIFIER:
      updatedSkill = {
        ...state[action?.payload.skillName],
        miscModifier: action?.payload.miscModifier,
      }
      break

    case UPDATE_SKILL_RANKS:
      if (action?.payload.ranks < 0) {
        // TODO: don't accept ranks > character level
        console.warn('Ranks must be >= 0')
      } else {
        updatedSkill = {
          ...state[action?.payload.skillName],
          ranks: action?.payload.ranks,
        }
      }

      break

    case UPDATE_SKILL_IS_CLASS_SKILL:
      updatedSkill = {
        ...state[action?.payload.skillName],
        isClassSkill: action?.payload.isClassSkill,
      }

      break
  }

  // @ts-ignore updatedSkill is used before it is assigned
  if (updatedSkill && action) {
    updateTotalSkillBonus(updatedSkill)

    return {
      ...state,
      [action?.payload.skillName]: updatedSkill,
    }
  } else {
    return state
  }
}
