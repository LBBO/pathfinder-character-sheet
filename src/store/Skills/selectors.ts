import { quickSkillDefinitions, Skill, SkillState } from './types'
import { createSelector } from 'reselect'
import { getAbilityModifiers } from '../Abilities/selectors'
import { abilityName } from '../Abilities/types'
import { RootState } from '../root-reducer'
import { getTotalArmorCheckPenalty } from '../Inventory/getters'

export const getSkills = (state: RootState) => state.skills
export type TotalSkillBonuses = {
  [k in keyof SkillState]: number
}

export const computeTotalSkillBonus = (
  skill: Skill,
  abilityModifier: number,
): number => {
  return (
    abilityModifier +
    skill.ranks +
    skill.miscModifier +
    (skill.isClassSkill && skill.ranks >= 1 ? 3 : 0)
  )
}

export const getTotalSkillBonuses = createSelector(
  [getSkills, getAbilityModifiers, getTotalArmorCheckPenalty],
  (skills, abilityModifiers, totalCheckPenalty) => {
    const skillNames = Object.keys(skills) as Array<keyof SkillState>
    return skillNames.reduce((totalSkillsBonuses, currSkillName) => {
      const baseAbilityName = quickSkillDefinitions[currSkillName]
        .baseAbility as abilityName

      let abilityModifier = abilityModifiers[baseAbilityName]

      // Certain armor items can add a penalty to STR and DEX checks
      // See German pathfinder documentation:
      // http://prd.5footstep.de/Grundregelwerk/Ausruestung/Ruestungen/Ruestungseigenschaften
      if (baseAbilityName === 'strength' || baseAbilityName === 'dexterity') {
        abilityModifier -= totalCheckPenalty
      }

      return {
        ...totalSkillsBonuses,
        [currSkillName]: computeTotalSkillBonus(
          skills[currSkillName],
          abilityModifier,
        ),
      }
    }, {} as TotalSkillBonuses)
  },
)
