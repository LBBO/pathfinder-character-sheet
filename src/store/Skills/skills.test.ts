import { quickSkillDefinitions, Skill, SkillState } from './types'
import { createInitialState, SkillsReducer } from './reducers'
import { setAbilityScore } from '../Abilities/actions'
import { setIsSkillClassSkill, setSkillMiscMod, setSkillRanks } from './actions'
import { EmptyAction, RootReducer, RootState } from '../root-reducer'
import { abilityName } from '../Abilities/types'
import { computeTotalSkillBonus, getTotalSkillBonuses } from './selectors'
import { setCharacterSizeCategory } from '../CharacterMetaData/actions'
import { SizeCategory } from '../CharacterMetaData/Character'
import { addArmorItem } from '../Inventory/actions'

let initialRootState: RootState
let initialState: SkillState
const newModifierValue = 1

const setup = () => {
  initialRootState = RootReducer(undefined, EmptyAction)
  const abilityNames: abilityName[] = [
    'charisma',
    'constitution',
    'dexterity',
    'intelligence',
    'strength',
    'wisdom',
  ]

  abilityNames.forEach((abilityName) => {
    initialRootState = RootReducer(
      initialRootState,
      setAbilityScore(abilityName, 10 + newModifierValue * 2),
    )
  })

  initialState = createInitialState()
}

beforeEach(setup)
setup()

describe('total skill bonus calculator', () => {
  let skill: Skill
  let abilityModifier: number

  beforeEach(() => {
    skill = {
      isClassSkill: false,
      name: 'acrobatics',
      miscModifier: 0,
      ranks: 0,
    }
    abilityModifier = 0
  })

  it('should return 0 when every modifier is 0', () => {
    expect(computeTotalSkillBonus(skill, abilityModifier)).toBe(0)
  })

  it('should include ability mod, ranks and misc mod in calculation', () => {
    skill.ranks = 1
    abilityModifier = 3
    skill.miscModifier = -2

    expect(computeTotalSkillBonus(skill, abilityModifier)).toBe(2)
  })

  it('should add a bonus of 3 if a class skill has at least one rank', () => {
    skill.isClassSkill = true
    skill.ranks = 0

    expect(computeTotalSkillBonus(skill, abilityModifier)).toBe(0)

    skill.ranks = 1

    // ranks + bonus for class skill = 4
    expect(computeTotalSkillBonus(skill, abilityModifier)).toBe(4)
  })
})

describe('createInitialState', () => {
  it('should always create the same state when called with initial modifiers', () => {
    expect(createInitialState()).toMatchSnapshot()
  })
})

describe('SkillsReducer', () => {
  it('should return the correct initial state', () => {
    expect(SkillsReducer()).toMatchObject(initialState)
  })

  it('should update a skill correctly when a misc mod is changed', () => {
    const action = setSkillMiscMod('acrobatics', 2)
    const newState = SkillsReducer(initialState, action)
    expect(newState).toMatchObject({
      ...initialState,
      acrobatics: {
        ...initialState.acrobatics,
        miscModifier: 2,
      },
    })
  })

  it('should update a skill correctly when its ranks are changed', () => {
    const action = setSkillRanks('acrobatics', 2)
    const newState = SkillsReducer(initialState, action)
    expect(newState).toMatchObject({
      ...initialState,
      acrobatics: {
        ...initialState.acrobatics,
        ranks: 2,
      },
    })
  })

  it('should update a skill correctly when its class skill flag is changed', () => {
    const setClassSkillAction = setIsSkillClassSkill('acrobatics', true)
    const stateWithClassSkill = SkillsReducer(initialState, setClassSkillAction)
    expect(stateWithClassSkill).toMatchObject({
      ...initialState,
      acrobatics: {
        ...initialState.acrobatics,
        isClassSkill: true,
      },
    })

    const setRanksAction = setSkillRanks('acrobatics', 1)
    const stateWithRank = SkillsReducer(stateWithClassSkill, setRanksAction)
    expect(stateWithRank).toMatchObject({
      ...initialState,
      acrobatics: {
        ...initialState.acrobatics,
        isClassSkill: true,
        ranks: 1,
      },
    })
  })
})

describe('total skill bonus selector', () => {
  it('should update when an ability is changed', () => {
    const oldTotalBonuses = getTotalSkillBonuses(initialRootState)
    const action = setAbilityScore('dexterity', 14)
    const newState = RootReducer(initialRootState, action)
    const newTotalBonuses = getTotalSkillBonuses(newState)

    expect(oldTotalBonuses).not.toBe(newTotalBonuses)
    expect(quickSkillDefinitions.acrobatics.baseAbility).toBe('dexterity')
    expect(newTotalBonuses.acrobatics).toBe(2)
  })

  it('should update when a skill is changed', () => {
    initialRootState = RootReducer(
      initialRootState,
      setAbilityScore('dexterity', 10),
    )
    const oldTotalBonuses = getTotalSkillBonuses(initialRootState)
    const action = setSkillRanks('acrobatics', 1)
    const newState = RootReducer(initialRootState, action)
    const newTotalBonuses = getTotalSkillBonuses(newState)

    expect(oldTotalBonuses).not.toBe(newTotalBonuses)
    expect(newTotalBonuses.acrobatics).toBe(1)
  })

  it('should not update something else is changed', () => {
    const oldTotalBonuses = getTotalSkillBonuses(initialRootState)
    const action = setCharacterSizeCategory(SizeCategory.SMALL)
    const newState = RootReducer(initialRootState, action)
    const newTotalBonuses = getTotalSkillBonuses(newState)

    expect(oldTotalBonuses).toBe(newTotalBonuses)
  })

  it('should subtract total armor check penalty from DEX and STR skills', () => {
    let state = RootReducer(
      undefined,
      addArmorItem({
        name: '',
        type: '',
        properties: '',
        checkPenalty: 1,
      }),
    )
    state = RootReducer(state, setSkillRanks('acrobatics', 2))
    state = RootReducer(state, setSkillRanks('climb', 2))

    expect(getTotalSkillBonuses(state).acrobatics).toBe(1)
    expect(getTotalSkillBonuses(state).climb).toBe(1)
  })
})
