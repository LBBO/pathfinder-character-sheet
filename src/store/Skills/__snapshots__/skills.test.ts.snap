// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createInitialState should always create the same state when called with initial modifiers 1`] = `
Object {
  "acrobatics": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "acrobatics",
    "ranks": 0,
    "totalBonus": 0,
  },
  "appraise": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "appraise",
    "ranks": 0,
    "totalBonus": 0,
  },
  "bluff": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "bluff",
    "ranks": 0,
    "totalBonus": 0,
  },
  "climb": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "climb",
    "ranks": 0,
    "totalBonus": 0,
  },
  "craft": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "craft",
    "ranks": 0,
    "totalBonus": 0,
  },
  "diplomacy": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "diplomacy",
    "ranks": 0,
    "totalBonus": 0,
  },
  "disableDevice": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "disableDevice",
    "ranks": 0,
    "totalBonus": 0,
  },
  "disguise": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "disguise",
    "ranks": 0,
    "totalBonus": 0,
  },
  "escapeArtist": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "escapeArtist",
    "ranks": 0,
    "totalBonus": 0,
  },
  "fly": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "fly",
    "ranks": 0,
    "totalBonus": 0,
  },
  "handleAnimal": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "handleAnimal",
    "ranks": 0,
    "totalBonus": 0,
  },
  "heal": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "heal",
    "ranks": 0,
    "totalBonus": 0,
  },
  "intimidate": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "intimidate",
    "ranks": 0,
    "totalBonus": 0,
  },
  "knowledgeArcana": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "knowledgeArcana",
    "ranks": 0,
    "totalBonus": 0,
  },
  "knowledgeDungeoneering": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "knowledgeDungeoneering",
    "ranks": 0,
    "totalBonus": 0,
  },
  "knowledgeEngineering": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "knowledgeEngineering",
    "ranks": 0,
    "totalBonus": 0,
  },
  "knowledgeGeography": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "knowledgeGeography",
    "ranks": 0,
    "totalBonus": 0,
  },
  "knowledgeHistory": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "knowledgeHistory",
    "ranks": 0,
    "totalBonus": 0,
  },
  "knowledgeLocal": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "knowledgeLocal",
    "ranks": 0,
    "totalBonus": 0,
  },
  "knowledgeNature": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "knowledgeNature",
    "ranks": 0,
    "totalBonus": 0,
  },
  "knowledgeNobility": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "knowledgeNobility",
    "ranks": 0,
    "totalBonus": 0,
  },
  "knowledgePlanes": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "knowledgePlanes",
    "ranks": 0,
    "totalBonus": 0,
  },
  "knowledgeReligion": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "knowledgeReligion",
    "ranks": 0,
    "totalBonus": 0,
  },
  "linguistics": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "linguistics",
    "ranks": 0,
    "totalBonus": 0,
  },
  "perception": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "perception",
    "ranks": 0,
    "totalBonus": 0,
  },
  "perform": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "perform",
    "ranks": 0,
    "totalBonus": 0,
  },
  "profession": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "profession",
    "ranks": 0,
    "totalBonus": 0,
  },
  "ride": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "ride",
    "ranks": 0,
    "totalBonus": 0,
  },
  "senseMotive": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "senseMotive",
    "ranks": 0,
    "totalBonus": 0,
  },
  "sleightOfHand": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "sleightOfHand",
    "ranks": 0,
    "totalBonus": 0,
  },
  "spellcraft": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "spellcraft",
    "ranks": 0,
    "totalBonus": 0,
  },
  "stealth": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "stealth",
    "ranks": 0,
    "totalBonus": 0,
  },
  "survival": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "survival",
    "ranks": 0,
    "totalBonus": 0,
  },
  "swim": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "swim",
    "ranks": 0,
    "totalBonus": 0,
  },
  "useMagicDevice": Object {
    "abilityModifier": 0,
    "isClassSkill": false,
    "miscModifier": 0,
    "name": "useMagicDevice",
    "ranks": 0,
    "totalBonus": 0,
  },
}
`;
